engine/consensus/matching/engine.go

```diff
@@ -462,169 +458,163 @@ func (e *Engine) checkSealing() {
+		// blockID is the first block in its fork, which references an Execution
+		// Receipt with an Execution Result identical to result.
+		// (where blockID != result.BlockID)
+		assignment, err := e.assigner.Assign(result, result.BlockID)
+		if state.IsNoValidChildBlockError(err) {
+			continue
+		}
+		if err != nil {
+			return nil, fmt.Errorf("could not assign verifiers: %w", err)
+		}
+
+		// check that each chunk collects enough approvals
+		for _, chunk := range result.Chunks {
+			matched := e.matchChunk(result.ID(), chunk, assignment)
+			if !matched {
+				allChunksMatched = false
+				break
 			}
 		}
-		e.log.Info().
-			Int("count", requestedCount).
-			Msg("requested missing results")
+
+		// add the result to the results that should be sealed
+		if allChunksMatched {
+			e.log.Info().Msg("adding result with sufficient verification")
+			results = append(results, result)
+		}
 	}
 
 	return results, nil
 }
 
-func (e *Engine) sealResult(result *flow.ExecutionResult, final *flow.Header) error {
+// matchChunk checks that the number of ResultApprovals collected by a chunk
+// exceeds the required threshold.
+func (e *Engine) matchChunk(resultID flow.Identifier, chunk *flow.Chunk, assignment *chunks.Assignment) bool {
 
-	// check if we know the block the result pertains to
-	_, err := e.headersDB.ByBlockID(result.BlockID)
-	if errors.Is(err, storage.ErrNotFound) {
-		return errUnknownBlock
-	}
-	if err != nil {
-		return fmt.Errorf("could not check sealed header: %w", err)
-	}
+	// get all the chunk approvals from mempool
+	approvals := e.approvals.ByChunk(resultID, chunk.Index)
 
-	// get the previous result from our mempool or storage
-	previousID := result.PreviousResultID
-	previous, found := e.results.ByID(previousID)
-	if !found {
-		var err error
-		previous, err = e.resultsDB.ByID(previousID)
-		if errors.Is(err, storage.ErrNotFound) {
-			return errUnknownPrevious
-		}
-		if err != nil {
-			return fmt.Errorf("could not get previous result: %w", err)
+	// only keep approvals from assigned verifiers
+	var validApprovers flow.IdentifierList
+	for approverID := range approvals {
+		ok := chmodule.IsValidVerifer(assignment, chunk, approverID)
+		if ok {
+			validApprovers = append(validApprovers, approverID)
 		}
 	}
 
+	// skip counting approvals
+	// TODO:
+	//   * this is only here temporarily to ease the migration to new chunk
+	//     based sealing.
+	if !e.requireApprovals {
+		return true
+	}
+
+	// TODO:
+	//   * This is the happy path (requires just one approval per chunk).
+	//   * Full protocol should be +2/3 of all currently staked verifiers.
+	return len(validApprovers) > 0
+}
+
+// sealResult records a matched result in the result database, creates a seal
+// for the corresponding block and saves in the database and seal mempool for
+// further processing.
+func (e *Engine) sealResult(result *flow.ExecutionResult) error {
+
 	// store the result to make it persistent for later checks
```

Comment:
Yes, phase 1 does not implement fork-aware sealing yet. In phase 3, the matching engine will not use resultsDB to check if results are sealed or not. Instead, it will rely on the Finalizer, which is fork aware, to submit IncorporatedResults for sealing.